Argument of type "null | string" is not assignable to parameter of type "null | string"

**Bicep version**
0.12.1

**Describe the bug**
After upgrading to 0.12.1 we get the following error on conditional property values using the ternary operator. The code did work for at least 16 months before, but after the upgrade the compilation fails. Even the error message is confusing as the mentioned types match:

`Argument of type "null | string" is not assignable to parameter of type "null | string"`

We are getting this error for the following resource block:

![image](https://user-images.githubusercontent.com/1492319/199955691-f51f6f24-9507-4346-8b84-b7eb3c733b18.png)

We didn't change that code for over a year.

**To Reproduce**
See screenshot

**Additional context**
We compiled the exact same code with previous bicep versions and it works. So it seems this error is definitely caused by version 0.12.1

> @SeSeicht -- VS Code and the hosted agents are both capable of auto-upgrading. Can you confirm you are not running `v0.12` by checking `az bicep version` in your pipeline and/or checking the extension: <img alt="image" width="1170" src="https://user-images.githubusercontent.com/11078275/200019685-589c016d-3734-4130-a1f1-d3e49b5eb3ab.png">

```
2022-11-04T15:58:56.9203654Z /home/vsts/work/1/s/policies/assignments/policyAssignmentsModule/resourceGroup/deploy.bicep(74,20) : Error BCP070: Argument of type "null | object | object" is not assignable to parameter of type "Identity | null".
2022-11-04T15:58:56.9204601Z /home/vsts/work/1/s/policies/assignments/policyAssignmentsModule/resourceGroup/deploy.bicep(74,20) : Error BCP070: Argument of type "null | object | object" is not assignable to parameter of type "Identity | null".
2022-11-04T15:58:56.9207063Z /home/vsts/work/1/s/policies/assignments/policyAssignmentsModule/managementGroup/deploy.bicep(59,20) : Error BCP070: Argument of type "null | object | object" is not assignable to parameter of type "Identity | null".
2022-11-04T15:58:56.9207814Z /home/vsts/work/1/s/policies/assignments/policyAssignmentsModule/managementGroup/deploy.bicep(59,20) : Error BCP070: Argument of type "null | object | object" is not assignable to parameter of type "Identity | null".
2022-11-04T15:58:56.9209356Z /home/vsts/work/1/s/policies/assignments/PolicyAssignments_cx.bicep(105,31) : Error BCP104: The referenced module has errors.
2022-11-04T15:58:56.9209986Z /home/vsts/work/1/s/policies/assignments/policyAssignmentsModule/deploy.bicep(1,1) : Info Bicep Linter Configuration: Custom bicepconfig.json file found (/home/vsts/work/1/s/bicepconfig.json).
2022-11-04T15:58:56.9210620Z /home/vsts/work/1/s/policies/assignments/policyAssignmentsModule/deploy.bicep(79,28) : Error BCP104: The referenced module has errors.
2022-11-04T15:58:56.9211518Z /home/vsts/work/1/s/policies/assignments/policyAssignmentsModule/deploy.bicep(101,29) : Error BCP104: The referenced module has errors.
2022-11-04T15:58:56.9212075Z /home/vsts/work/1/s/policies/assignments/policyAssignmentsModule/deploy.bicep(123,28) : Error BCP104: The referenced module has errors.
2022-11-04T15:58:56.9212693Z /home/vsts/work/1/s/policies/exemptions/PolicyExemptions_cx.bicep(1,1) : Info Bicep Linter Configuration: Custom bicepconfig.json file found (/home/vsts/work/1/s/bicepconfig.json).
2022-11-04T15:58:56.9213385Z /home/vsts/work/1/s/policies/definitions/PolicyDefinitions_cx.bicep(1,1) : Info Bicep Linter Configuration: Custom bicepconfig.json file found (/home/vsts/work/1/s/bicepconfig.json).
2022-11-04T15:58:56.9214109Z /home/vsts/work/1/s/policies/assignments/policyAssignmentsModule/subscription/deploy.bicep(1,1) : Info Bicep Linter Configuration: Custom bicepconfig.json file found (/home/vsts/work/1/s/bicepconfig.json).
2022-11-04T15:58:56.9214873Z /home/vsts/work/1/s/policies/assignments/policyAssignmentsModule/subscription/deploy.bicep(59,20) : Error BCP070: Argument of type "null | object | object" is not assignable to parameter of type "Identity | null".
2022-11-04T15:58:56.9215619Z /home/vsts/work/1/s/policies/assignments/policyAssignmentsModule/subscription/deploy.bicep(59,20) : Error BCP070: Argument of type "null | object | object" is not assignable to parameter of type "Identity | null".
2022-11-04T15:58:56.9216313Z /home/vsts/work/1/s/policies/deploy.bicep(1,1) : Info Bicep Linter Configuration: Custom bicepconfig.json file found (/home/vsts/work/1/s/bicepconfig.json).
2022-11-04T15:58:56.9216849Z /home/vsts/work/1/s/policies/deploy.bicep(16,27) : Error BCP104: The referenced module has errors.
2022-11-04T15:58:56.9996289Z [33;1mVERBOSE: Invoke task with[0m
2022-11-04T15:58:57.0046898Z [33;1mVERBOSE: {
2022-11-04T15:58:57.0047307Z   "subscriptionId": "$(subscriptionId)",
2022-11-04T15:58:57.0047656Z   "templateFilePath": "/home/vsts/work/1/s/policies/deploy.bicep",
2022-11-04T15:58:57.0047968Z   "additionalParameters": {},
2022-11-04T15:58:57.0048339Z   "parameterFilePath": "/home/vsts/work/1/s/policies/.parameters/dev.weu.parameters.json",
2022-11-04T15:58:57.0048712Z   "location": "westeurope",
2022-11-04T15:58:57.0049127Z   "managementGroupId": "dev-mg-cx",
2022-11-04T15:58:57.0049443Z   "resourceGroupName": "$(resourceGroupName)"
2022-11-04T15:58:57.0049690Z }
2022-11-04T15:58:57.0049980Z [0m
2022-11-04T15:58:57.0336886Z [33;1mVERBOSE: Determined deployment scope [managementGroup][0m
2022-11-04T15:58:57.0439327Z [33;1mVERBOSE: Testing with deployment name [policies-20221104T1511570378Z][0m
2022-11-04T15:58:57.0457514Z [33;1mVERBOSE: Performing the operation "Test" on target "Management group level deployment".[0m
2022-11-04T15:59:03.9792490Z [33;1mVERBOSE: Using Bicep v0.11.1[0m
...
2022-11-04T15:59:31.0609020Z [33;1mVERBOSE: 15:59:31 - Template is valid.[0m
2022-11-04T15:59:31.1892187Z [33;1mVERBOSE: Template is valid[0m
2022-11-04T15:59:31.6351996Z ##[error]Script failed with exit code: 1
2022-11-04T15:59:31.6373994Z [command]/usr/bin/az account clear
```

@alex-frankel  this is a shortend output from an ADO pipeline with ubuntu-latest which says that its using bicep v0.11.1

--- a/src/Bicep.Core.IntegrationTests/ScenarioTests.cs
+++ b/src/Bicep.Core.IntegrationTests/ScenarioTests.cs
@@ -4018,7 +4018,8 @@ resource functionApp 'Microsoft.Web/sites@2022-03-01' = {
 
             result.Should().NotHaveAnyDiagnostics();
         }
-        
+
+        /// <summary>
         /// https://github.com/Azure/bicep/issues/8884
         /// </summary>
         [TestMethod]

--- a/src/Bicep.Core.IntegrationTests/ScenarioTests.cs
+++ b/src/Bicep.Core.IntegrationTests/ScenarioTests.cs
@@ -3971,5 +3971,52 @@ resource queueAuthorizationRules 'Microsoft.ServiceBus/namespaces/queues/authori
             result.Template.Should().HaveValueAtPath("$.resources[2].name", "[format('{0}/{1}/{2}', variables('Names')[0], variables('Service_Bus_Queues')[copyIndex()], 'Listen')]");
             result.Template.Should().HaveValueAtPath("$.resources[2].dependsOn", new JArray("[resourceId('Microsoft.ServiceBus/namespaces/queues', variables('Names')[0], variables('Service_Bus_Queues')[copyIndex()])]"));
         }
+
+        /// <summary>
+        /// https://github.com/Azure/bicep/issues/8890
+        /// </summary>
+        [TestMethod]
+        public void Test_Issue8890()
+        {
+            var result = CompilationHelper.Compile(@"
+param location string = resourceGroup().location
+
+@description('Optional. Enables system assigned managed identity on the resource.')
+param systemAssignedIdentity bool = false
+
+@description('Optional. The ID(s) to assign to the resource.')
+param userAssignedIdentities object = {}
+
+var identityType = systemAssignedIdentity ? (!empty(userAssignedIdentities) ? 'SystemAssigned,UserAssigned' : 'SystemAssigned') : (!empty(userAssignedIdentities) ? 'UserAssigned' : 'None')
+
+var identity = identityType != 'None' ? {
+  type: identityType
+  userAssignedIdentities: !empty(userAssignedIdentities) ? userAssignedIdentities : null
+} : null
+
+resource vm 'Microsoft.Compute/virtualMachines@2021-07-01' = {
+  name: 'name'
+  location: location
+  identity: identity
+}
+
+param usePython bool
+
+resource functionApp 'Microsoft.Web/sites@2022-03-01' = {
+  name: 'fa'
+  location: location
+  kind: 'functionApp'
+  identity: identity
+  properties: {
+    siteConfig: {
+      pythonVersion: usePython ? '~3.10' : null
+      nodeVersion: !usePython ? '18' : null
+    }
+  }
+}
+");
+
+            result.Should().NotHaveAnyDiagnostics();
+        }
     }
 }

--- a/src/Bicep.Core/TypeSystem/TypeValidator.cs
+++ b/src/Bicep.Core/TypeSystem/TypeValidator.cs
@@ -446,22 +446,21 @@ namespace Bicep.Core.TypeSystem
 
                 var candidateDiagnostics = ToListDiagnosticWriter.Create();
                 var candidateCollector = new TypeValidator(typeManager, binder, candidateDiagnostics);
-                var narrowed = candidateCollector.NarrowUnionTypeForSingleExpressionType(config, expression, candidateExpressionType.Type, targetType);
+                var narrowed = candidateCollector.NarrowUnionType(config, expression, candidateExpressionType.Type, targetType);
                 candidacyEvaluations.Add(new(candidateExpressionType, narrowed, candidateDiagnostics.GetDiagnostics()));
             }
 
             // report all informational and warning diagnostics
-            diagnosticWriter.WriteMultiple(candidacyEvaluations.SelectMany(e => e.Diagnostics.Where(d => d.Level != DiagnosticLevel.Error)));
+            diagnosticWriter.WriteMultiple(candidacyEvaluations.SelectMany(e => e.Diagnostics));
 
             var errors = candidacyEvaluations.SelectMany(c =>
             {
-                var reportedErrors = c.Diagnostics.OfType<ErrorDiagnostic>();
-                if (c.Narrowed is null || reportedErrors.Any())
+                if (c.NarrowedType is null || c.Errors.Any())
                 {
-                    reportedErrors = reportedErrors.Append(DiagnosticBuilder.ForPosition(expression).ArgumentTypeMismatch(c.Type.Type, targetType));
+                    return c.Errors.Append(DiagnosticBuilder.ForPosition(expression).ArgumentTypeMismatch(c.UnionTypeMemberEvaluated.Type, targetType));
                 }
 
-                return reportedErrors;
+                return c.Errors;
             });
 
             // If *any* variant of the expression type is not assignable, the expression type is not assignable
@@ -470,7 +469,7 @@ namespace Bicep.Core.TypeSystem
                 return ErrorType.Create(errors);
             }
 
-            return TypeHelper.CreateTypeUnion(candidacyEvaluations.Select(c => c.Narrowed!));
+            return TypeHelper.CreateTypeUnion(candidacyEvaluations.Select(c => c.NarrowedType!));
         }
 
         private TypeSymbol NarrowUnionTypeForSingleExpressionType(TypeValidatorConfig config, SyntaxBase expression, TypeSymbol expressionType, UnionType targetType)
@@ -492,7 +491,7 @@ namespace Bicep.Core.TypeSystem
             }
 
             var viableCandidates = candidacyEvaluations
-                .Select(c => c.Narrowed is {} Narrowed && !c.HasErrors ? new ViableTypeCandidate(Narrowed, c.Diagnostics) : null)
+                .Select(c => c.NarrowedType is {} Narrowed && !c.Errors.Any() ? new ViableTypeCandidate(Narrowed, c.Diagnostics) : null)
                 .WhereNotNull()
                 .ToImmutableArray();
 
@@ -511,9 +510,57 @@ namespace Bicep.Core.TypeSystem
             return ErrorType.Create(DiagnosticBuilder.ForPosition(expression).ArgumentTypeMismatch(expressionType, targetType));
         }
 
-        private record UnionTypeMemberViabilityInfo(ITypeReference Type, TypeSymbol? Narrowed, IEnumerable<IDiagnostic> Diagnostics)
+        private record UnionTypeMemberViabilityInfo
         {
-            public bool HasErrors => Diagnostics.OfType<ErrorDiagnostic>().Any();
+            internal UnionTypeMemberViabilityInfo(ITypeReference unionTypeMemberEvaluated, TypeSymbol? narrowedType, IEnumerable<IDiagnostic> diagnostics)
+            {
+                UnionTypeMemberEvaluated = unionTypeMemberEvaluated;
+                NarrowedType = narrowedType;
+
+                List<IDiagnostic> nonErrorDiagnostics = new();
+                List<ErrorDiagnostic> errorDiagnostics = new();
+
+                foreach (var diagnostic in diagnostics)
+                {
+                    if (diagnostic.Level == DiagnosticLevel.Error)
+                    {
+                        errorDiagnostics.Add(AsErrorDiagnostic(diagnostic));
+                    }
+                    else
+                    {
+                        nonErrorDiagnostics.Add(diagnostic);
+                    }
+                }
+
+                Diagnostics = nonErrorDiagnostics;
+                Errors = errorDiagnostics;
+            }
+
+            /// <summary>
+            /// The type being checked for assignability and narrowing
+            /// </summary>
+            public ITypeReference UnionTypeMemberEvaluated { get; }
+
+            /// <summary>
+            /// The narrowed type. Will be null if the type is unassignable
+            /// </summary>
+            public TypeSymbol? NarrowedType { get; }
+
+            /// <summary>
+            /// Any warning or informational diagnostics raised during type narrowing
+            /// </summary>
+            public IReadOnlyList<IDiagnostic> Diagnostics { get; }
+
+            /// <summary>
+            /// Any error-level diagnostics raised during type narrowing
+            /// </summary>
+            public IReadOnlyList<ErrorDiagnostic> Errors { get; }
+
+            private static ErrorDiagnostic AsErrorDiagnostic(IDiagnostic diagnostic) => diagnostic switch
+            {
+                ErrorDiagnostic errorDiagnostic => errorDiagnostic,
+                _ => new(diagnostic.Span, diagnostic.Code, diagnostic.Message, diagnostic.Uri, diagnostic.Styling),
+            };
         }
 
         private record ViableTypeCandidate(TypeSymbol Type, IEnumerable<IDiagnostic> Diagnostics);

--- a/src/Bicep.Core/TypeSystem/TypeValidator.cs
+++ b/src/Bicep.Core/TypeSystem/TypeValidator.cs
@@ -215,9 +215,10 @@ namespace Bicep.Core.TypeSystem
         }
 
         private TypeSymbol NarrowType(TypeValidatorConfig config, SyntaxBase expression, TypeSymbol targetType)
-        {
-            var expressionType = typeManager.GetTypeInfo(expression);
+            => NarrowType(config, expression, typeManager.GetTypeInfo(expression), targetType);
 
+        private TypeSymbol NarrowType(TypeValidatorConfig config, SyntaxBase expression, TypeSymbol expressionType, TypeSymbol targetType)
+        {
             if (config.DisallowAny && expressionType is AnyType)
             {
                 // certain properties such as scope, parent, dependsOn do not allow values of "any" type
@@ -423,42 +424,86 @@ namespace Bicep.Core.TypeSystem
             return targetType;
         }
 
+
         private TypeSymbol NarrowUnionType(TypeValidatorConfig config, SyntaxBase expression, TypeSymbol expressionType, UnionType targetType)
+            => expressionType switch
+            {
+                UnionType expressionUnion => NarrowUnionTypeForUnionExpressionType(config, expression, expressionUnion, targetType),
+                _ => NarrowUnionTypeForSingleExpressionType(config, expression, expressionType, targetType),
+            };
+
+        private TypeSymbol NarrowUnionTypeForUnionExpressionType(TypeValidatorConfig config, SyntaxBase expression, UnionType expressionType, UnionType targetType)
         {
             List<UnionTypeMemberViabilityInfo> candidacyEvaluations = new();
 
-            foreach (var candidateType in targetType.Members)
+            foreach (var candidateExpressionType in expressionType.Members)
             {
-                if (!AreTypesAssignable(expressionType, candidateType.Type))
+                if (!AreTypesAssignable(candidateExpressionType.Type, targetType))
                 {
-                    candidacyEvaluations.Add(new(candidateType, null, ImmutableArray<IDiagnostic>.Empty));
+                    candidacyEvaluations.Add(new(candidateExpressionType, null, ImmutableArray<IDiagnostic>.Empty));
                     continue;
                 }
 
                 var candidateDiagnostics = ToListDiagnosticWriter.Create();
                 var candidateCollector = new TypeValidator(typeManager, binder, candidateDiagnostics);
-                var narrowed = candidateCollector.NarrowType(config, expression, candidateType.Type);
+                var narrowed = candidateCollector.NarrowUnionTypeForSingleExpressionType(config, expression, candidateExpressionType.Type, targetType);
+                candidacyEvaluations.Add(new(candidateExpressionType, narrowed, candidateDiagnostics.GetDiagnostics()));
+            }
 
-                // if we have a perfect match, skip checking the rest of the union members
-                if (!candidateDiagnostics.GetDiagnostics().Any())
+            // report all informational and warning diagnostics
+            diagnosticWriter.WriteMultiple(candidacyEvaluations.SelectMany(e => e.Diagnostics.Where(d => d.Level != DiagnosticLevel.Error)));
+
+            var errors = candidacyEvaluations.SelectMany(c =>
+            {
+                var reportedErrors = c.Diagnostics.OfType<ErrorDiagnostic>();
+                if (c.Narrowed is null || reportedErrors.Any())
                 {
-                    return narrowed;
+                    reportedErrors = reportedErrors.Append(DiagnosticBuilder.ForPosition(expression).ArgumentTypeMismatch(c.Type.Type, targetType));
                 }
 
-                candidacyEvaluations.Add(new(candidateType, narrowed, candidateDiagnostics.GetDiagnostics()));
+                return reportedErrors;
+            });
+
+            // If *any* variant of the expression type is not assignable, the expression type is not assignable
+            if (errors.Any())
+            {
+                return ErrorType.Create(errors);
+            }
+
+            return TypeHelper.CreateTypeUnion(candidacyEvaluations.Select(c => c.Narrowed!));
+        }
+
+        private TypeSymbol NarrowUnionTypeForSingleExpressionType(TypeValidatorConfig config, SyntaxBase expression, TypeSymbol expressionType, UnionType targetType)
+        {
+            List<UnionTypeMemberViabilityInfo> candidacyEvaluations = new();
+
+            foreach (var candidateTargetType in targetType.Members)
+            {
+                if (!AreTypesAssignable(expressionType, candidateTargetType.Type))
+                {
+                    candidacyEvaluations.Add(new(candidateTargetType, null, ImmutableArray<IDiagnostic>.Empty));
+                    continue;
+                }
+
+                var candidateDiagnostics = ToListDiagnosticWriter.Create();
+                var candidateCollector = new TypeValidator(typeManager, binder, candidateDiagnostics);
+                var narrowed = candidateCollector.NarrowType(config, expression, expressionType, candidateTargetType.Type);
+                candidacyEvaluations.Add(new(candidateTargetType, narrowed, candidateDiagnostics.GetDiagnostics()));
             }
 
             var viableCandidates = candidacyEvaluations
-                .Select(c => c.Narrowed is {} Narrowed && !c.Diagnostics.OfType<ErrorDiagnostic>().Any() ? new ViableTypeCandidate(Narrowed, c.Diagnostics) : null)
+                .Select(c => c.Narrowed is {} Narrowed && !c.HasErrors ? new ViableTypeCandidate(Narrowed, c.Diagnostics) : null)
                 .WhereNotNull()
                 .ToImmutableArray();
 
             if (viableCandidates.Any())
             {
-                // It's unclear what we should do with warning and informational diagnostics if there's no path that's assignable without issue.
-                // Should we report only diagnostics raised by every viable candidate? Or report only those raised by a single candidate?
-                // Erring on the side of caution here and just reporting them all.
-                diagnosticWriter.WriteMultiple(viableCandidates.SelectMany(c => c.Diagnostics));
+                // It's unclear what we should do with warning and informational diagnostics. Should we report only the intersection of diagnostics raised by every viable candidate?
+                // Erring on the side of caution here and just reporting them all unless there are one or more candidates that are assignable w/o warning.
+                if (viableCandidates.All(c => c.Diagnostics.Any()))
+                {
+                    diagnosticWriter.WriteMultiple(viableCandidates.SelectMany(c => c.Diagnostics));
+                }
 
                 return TypeHelper.CreateTypeUnion(viableCandidates.Select(c => c.Type));
             }
@@ -466,7 +511,10 @@ namespace Bicep.Core.TypeSystem
             return ErrorType.Create(DiagnosticBuilder.ForPosition(expression).ArgumentTypeMismatch(expressionType, targetType));
         }
 
-        private record UnionTypeMemberViabilityInfo(ITypeReference Type, TypeSymbol? Narrowed, IEnumerable<IDiagnostic> Diagnostics);
+        private record UnionTypeMemberViabilityInfo(ITypeReference Type, TypeSymbol? Narrowed, IEnumerable<IDiagnostic> Diagnostics)
+        {
+            public bool HasErrors => Diagnostics.OfType<ErrorDiagnostic>().Any();
+        }
 
         private record ViableTypeCandidate(TypeSymbol Type, IEnumerable<IDiagnostic> Diagnostics);
 

